### PR TITLE
Make the lexical Env of FunctionEmitter more robust.

### PR DESCRIPTION
**Pull request by sjrd**
The lexical `Env` of `FunctionEmitter` was previously not always complete. It was not a big deal, because it is only used to avoid spilling expressions into temporaris if they are known to be immutable.

However, handling class captures in nested JS classes will require the `Env` to be complete, so that we can distinguish local variables from class captures. It is also internally nicer to have complete `Env`s.

### CODE REVIEW

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -957,12 +964,11 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
 
               case _ =>
                 val temp = newSyntheticVar()
-                val newEnv = env.withDef(temp, arg.tpe, false)
-                innerEnv = newEnv
                 val computeTemp = pushLhsInto(
                     Lhs.VarDef(temp, arg.tpe, mutable = false), arg,
                     Set.empty)
                 computeTemp +=: extractedStatements
+                innerEnv = innerEnv.withDef(temp, arg.tpe, false)
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala]: Write `mutable = false`? Also a couple of occurrences below.

**Comment by sjrd** [comment]
> Wow. I didn't know this is actually incomplete :)

Me neither. I discovered that through weird wrong codegen in the nested JS classes branch.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala (hunk 1 above)
Comment (sjrd) [reply tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala]: I cannot. This is an *overload* pimped through an implicit class, and apparently that prevents scalac from understanding the named param as such.

Comment (gzm0) [reply tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala]: Blargh :P Ok

**Pull request merged**